binary_search_tree/binary_search_tree.py

```diff
@@ -0,0 +1,222 @@
+                                    to_replace = to_replace.left
+                                parent_node.left.data = to_replace.data
+                                parent_to_replace.left = to_replace.right
+                        return
+                    else:
+                        iterator = iterator.left
+                else:
+                    if iterator.right is None:
+                        raise KeyError('Node does not found')
+                    elif iterator.right.data == data:
+                        node = iterator.right
+                        parent_node = iterator
+                        if node.right is None and node.left is None:
+                            # if node has no child
+                            parent_node.right = None
+                        elif node.right is None:
+                            # if node only has left child
+                            parent_node.right = node.left
+                        elif node.left is None:
+                            # if node only has right child
+                            parent_node.right = node.right
+                        else:
+                            # if node have both chidren
+                            if node.right.left is None:
+                                node.right.left = node.left
+                                parent_node.right = node.right
+                            else:
+                                to_replace = node.right
+                                parent_to_replace = node
+                                while to_replace.left is not None:
+                                    parent_to_replace = to_replace
+                                    to_replace = to_replace.left
+                                parent_node.right.data = to_replace.data
+                                parent_to_replace.left = to_replace.right
+                        return
+                    else:
+                        iterator = iterator.right
+        return
+
+    def search_node(self, data):
+        """
+        search element in binary tree
+        Time Complexity: O(Logn)average case and O(n) in worst case
+        :param data: search data in bst
+        :return : return True if node with value data exist else return False
+        """
+        if self.root.data == data:
+            return True
+        else:
+            iterator = self.root
+            while iterator is not None:
+                if iterator.data == data:
+                    return True
+                elif iterator.data > data:
+                    iterator = iterator.left
+                else:
+                    iterator = iterator.right
+            return False
+
+    def print_preorder(self, node):
+        """
+        Print all element of binary tree in preorder using recurtion
+        """
+        if node is None:
+            return
+        elif node.left is None and node.right is None:
+            print(node.data, end=' ')
```

Comment:
Nothing just try to print every element of tree in single line...but I changed it now.